### PR TITLE
[issue 45] Add nsp security check

### LINKED ISSUE #45: Add NSP security check
Can basically copy/paste https://github.com/brave/browser-laptop/pull/205

### PR DESCRIPTION
**Pull request by therealklanni**
```
(+) 5 vulnerabilities found

│               │ Regular Expression Denial of Service
│ Name          │ hawk
│ Installed     │ 3.1.2
│ Vulnerable    │ <4.1.1
│ Patched       │ >=4.1.1
│ Path          │ vault@0.0.5 > babel-cli@6.4.5 > request@2.67.0 > hawk@3.1.2
│ More Info     │ https://nodesecurity.io/advisories/77

│               │ Regular Expression Denial of Service
│ Name          │ hawk
│ Installed     │ 3.1.2
│ Vulnerable    │ <4.1.1
│ Patched       │ >=4.1.1
│ Path          │ vault@0.0.5 > request@2.67.0 > hawk@3.1.2
│ More Info     │ https://nodesecurity.io/advisories/77

│               │ Denial-of-Service Extended Event Loop Blocking
│ Name          │ qs
│ Installed     │ 0.6.6
│ Vulnerable    │ <1.0.0
│ Patched       │ >= 1.x
│ Path          │ vault@0.0.5 > bitgo@0.11.64 > chain-node@0.0.17 > request@2.36.0 > qs@0.6.6
│ More Info     │ https://nodesecurity.io/advisories/28


│               │ Denial-of-Service Memory Exhaustion
│ Name          │ qs
│ Installed     │ 0.6.6
│ Vulnerable    │ <1.0.0
│ Patched       │ >= 1.x
│ Path          │ vault@0.0.5 > bitgo@0.11.64 > chain-node@0.0.17 > request@2.36.0 > qs@0.6.6
│ More Info     │ https://nodesecurity.io/advisories/29

│               │ Regular Expression Denial of Service
│ Name          │ hawk
│ Installed     │ 1.0.0
│ Vulnerable    │ <4.1.1
│ Patched       │ >=4.1.1
│ Path          │ vault@0.0.5 > bitgo@0.11.64 > chain-node@0.0.17 > request@2.36.0 > hawk@1.0.0
│ More Info     │ https://nodesecurity.io/advisories/77
```

Unfortunately, I don't recommend adding an exception for any of these packages. We should try to find a solution. The best course of action is probably to notify the maintainers (that aren't already aware).

Tracking issues:
[BitGo](/BitGo/BitGoJS) [#20](/BitGo/BitGoJS/issues/20)
[chain-node](/chain-engineering/chain-node) [#27](/chain-engineering/chain-node/issues/27)
[request](/request/request) [#2020](/request/request/issues/2020)

Closes #45

### CODE REVIEW

**Comment by mrose17** [comment]
@therealklanni - i think the order of fixing is: request, bitgo, and then vault. since it wouldn't be _optimal_ to merge a PR that breaks travis(!!), let me do some forking and testing, etc., and then talk to the repo owners... updates soon!

**Comment by therealklanni** [comment]
Yep, absolutely don't want this merged yet until resolved! :

**Comment by mrose17** [comment]
thanks. i am working my way down the dependencies now...

**Comment by therealklanni** [comment]
Here are the open issues I know of. Apparently @remy fixed hawk@3.1.3 which could resolve this for the current version of request here, without needing to wait for request to update (which was a blocker), so that's good news.

[BitGo](/BitGo/BitGoJS) [#20](/BitGo/BitGoJS/issues/20)
[chain-node](/chain-engineering/chain-node) [#27](/chain-engineering/chain-node/issues/27)
[request](/request/request) [#2020](/request/request/issues/2020)

**Comment by remy** [comment]
Since you pulled me in (and yep, hawk is fixed), I've been part of the team building https://snyk.io which gives you similar vuln reporting (and can/should be used as part of your CI) but also has guided remediation via updates and hosted patches. I'd definitely recommend checking it out.

**Comment by remy** [comment]
Oh. I should add, there's also the ability to create a policy file via the guided remediation, so if there's vulns you need to ignore for any specific reason, that is taken into account when you run `snyk test` in your project.

**Comment by mrose17** [comment]
keeping the issue -- https://github.com/brave/vault/tree/issue-41 -- but not using this fix!

File: package.json [hunk 1]
```diff
@@ -4,6 +4,8 @@
   "description": "A personal data store for holding high-value user behavior with high privacy.",
   "main": "index.js",
   "scripts": {
+    "checks": "npm run check-security",
+    "check-security": "nsp check",
```
Comment (mrose17) [inline package.json]: any chance we could make this "test" and "test-security" ???

**Comment by mrose17** [comment]
to re-iterate from https://github.com/brave/vault-client/pull/6

Should not be merged until the following issues are resolved

Tracking issues:
[BitGo](/BitGo/BitGoJS) [#20](http://github.com/BitGo/BitGoJS/issues/20)
[chain-node](/chain-engineering/chain-node) [#27](http://github.com/chain-engineering/chain-node/issues/27)
[request](/request/request) [#2020](http://github.com/request/request/issues/2020)

**Comment by mrose17** [comment]
fixed in https://github.com/brave/vault/commit/09efc74f2eab7300aba509dd1dbd7df702688e8a

**Pull request closed without merging**